src/hotspot/share/prims/jvmtiRawMonitor.cpp

```diff
@@ -306,75 +309,37 @@ void JvmtiRawMonitor::simple_notify(Thread* self, bool all) {
   return;
 }
 
-// Any JavaThread will enter here with state _thread_blocked unless we
-// are in single-threaded mode during startup.
-void JvmtiRawMonitor::raw_enter(Thread* self) {
-  void* contended;
-  JavaThread* jt = NULL;
-  // don't enter raw monitor if thread is being externally suspended, it will
-  // surprise the suspender if a "suspended" thread can still enter monitor
-  if (self->is_Java_thread()) {
-    jt = self->as_Java_thread();
-    while (true) {
-      // To pause suspend requests while in blocked we must block handshakes.
-      jt->handshake_state()->lock();
-      // Suspend request flag can only be set in handshakes.
-      // By blocking handshakes, suspend request flag cannot change its value.
-      if (!jt->handshake_state()->is_suspended()) {
-        contended = Atomic::cmpxchg(&_owner, (Thread*)NULL, jt);
-        jt->handshake_state()->unlock();
-        break;
-      }
-      jt->handshake_state()->unlock();
-
-      // We may only be in states other than _thread_blocked when we are
-      // in single-threaded mode during startup.
-      guarantee(jt->thread_state() == _thread_blocked, "invariant");
-
-      jt->set_thread_state_fence(_thread_blocked_trans);
-      SafepointMechanism::process_if_requested(jt);
-      // We should transition to thread_in_vm and then to thread_in_vm_trans,
-      // but those are always treated the same as _thread_blocked_trans.
-      jt->set_thread_state(_thread_blocked);
-    }
-  } else {
-    contended = Atomic::cmpxchg(&_owner, (Thread*)NULL, self);
-  }
+void JvmtiRawMonitor::ExitOnSuspend::operator()(JavaThread* current) {
+  // We must exit the monitor in case of a safepoint.
+  _rm->simple_exit(current);
+  _rm_exit = true;
+}
 
-  if (contended == self) {
+// JavaThreads will enter here with state _thread_in_native.
+void JvmtiRawMonitor::raw_enter(Thread* self) {
+  // TODO Atomic::load on _owner field
```

Comment:
We have ~15 plain loads of _owner, I thought it would be nice to fix all at the same time and I did not want to clobber this with that also.
I can do a follow-up and change them.